src/cis2/types.rs

```diff
@@ -0,0 +1,645 @@
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Burning an amount of tokens
+    #[display(
+        fmt = "Burned token with ID {} for {}",
+        token_id,
+        "address_display(owner)"
+    )]
+    Burn {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Add/Remove an address as operator for some other address.
+    #[display(
+        fmt = "{} {} as operator for {}",
+        update,
+        "address_display(operator)",
+        "address_display(owner)"
+    )]
+    UpdateOperator {
+        update:   OperatorUpdate,
+        owner:    Address,
+        operator: Address,
+    },
+    /// Provide an URL with the metadata for a certain token.
+    #[display(
+        fmt = "Added metadata url {} ({}) for token with ID {}",
+        "metadata_url.url",
+        "display_hash(metadata_url.hash)",
+        token_id
+    )]
+    TokenMetadata {
+        token_id:     TokenIdVec,
+        metadata_url: MetadataUrl,
+    },
+    /// Custom event outside of the CIS2 specification.
+    #[display(fmt = "Unknown event: Event is not part of CIS2 specification")]
+    Unknown,
+}
+
+fn display_hash(hash_opt: Option<Hash>) -> String {
+    if let Some(hash) = hash_opt {
+        format!("with hash {}", hex::encode(hash))
+    } else {
+        "without hash".to_string()
+    }
+}
+
+/// Deserialize the contract events as according to the CIS2 specification.
+impl Deserial for Event {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = u8::deserial(source)?;
+        match discriminant {
+            0 => Ok(Event::Transfer {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                from:     Address::deserial(source)?,
+                to:       Address::deserial(source)?,
+            }),
+            1 => Ok(Event::Mint {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            2 => Ok(Event::Burn {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            3 => Ok(Event::UpdateOperator {
+                update:   OperatorUpdate::deserial(source)?,
+                owner:    Address::deserial(source)?,
+                operator: Address::deserial(source)?,
+            }),
+            4 => Ok(Event::TokenMetadata {
+                token_id:     TokenIdVec::deserial(source)?,
+                metadata_url: MetadataUrl::deserial(source)?,
+            }),
+            _ => Ok(Event::Unknown),
+        }
+    }
+}
+
+/// Display the Address using either the display for account address or contract
+/// address.
+fn address_display(a: &Address) -> String {
```

Comment:
```suggestion
fn display_address(a: &Address) -> String {
```

In general for this (and hash below) it would be better to avoid allocating strings if possible. But it likely does not matter at the moment.

Reply:
Ideally we would implement Display for Address in concordium-contracts-common instead of this function.

Reply:
I have changed the function naming as you suggested.